syntax = "proto3";

message Request {
  string inputDataLocation = 1;
  int32 numberOfMappers = 2;
  int32 numberOfReducers = 3;
  string outputDataLocation = 4;
}

message Response {
  bool success = 1;
}

message RequestMapper {
  string inputDataLocation = 1;
  repeated Shard shards = 2;
}

message Shard {
  repeated string files = 1;
}

message RequestReducer {
  int32 numberOfReducers = 1;
  string intermediateOutputDataLocation = 2;
  string finalOutputDataLocation = 3;
}

message Empty {
}

service Master {
  rpc wordCount (Request) returns (Response) {}
}

service Mapper {
  rpc map(RequestMapper) returns (Response) {}
}

service Reducer {
  rpc reduce(RequestReducer) returns (Response) {}
}
